src/wattsi.pas

```diff
@@ -681,11 +686,61 @@    TCrossReferences = record
                            Scratch.Append($002E);
                         Scratch.Append(IntToStr(CurrentSectionNumber[CurrentHeadingRank]));
                      end;
+                     if (Variant = vDEV) then
+                     begin
+                        ID := Element.GetAttribute('id').AsString;
+                        SectionNumber := Scratch.AsString;
+                        HeadingText := DelSpace1(StringReplace(StringReplace(Element.TextContent.AsString, #$0A, ' ', [rfReplaceAll]), #$22, '\' + #$22, [rfReplaceAll]));
+                        if (not IsFirstSearchIndexItem) then
+                           Write(SearchIndexJsonFile, ',');
+                        IsFirstSearchIndexItem := False;
+                        Write(SearchIndexJsonFile, '{');
+                        Write(SearchIndexJsonFile, '"url":"' + SplitFilename + '.html#' + ID + '",');
+                        Write(SearchIndexJsonFile, '"text":"' + HeadingText + '",');
+                        if (LastDelimiter('.', SectionNumber) <> 0) then
+                        begin
+                           ParentSectionNumber := Copy(SectionNumber, 1, LastDelimiter('.', SectionNumber) - 1);
+                           ParentHeadingText := HeadingTextBySectionNumber[ParentSectionNumber];
```

Comment:
ParentHeadingText := HeadingTextBySectionNumber[ParentSectionNumber];

> This was the part I was really dreading.

I still feel like there might be some more-clever way to do this than building a hash table of every single heading in the document, but regardless, doing it this way doesn’t turn out to be costly in practice. so I’ll quit worrying :)